### PR TITLE
Change logic of L4 Strong Session Affinity

### PR DESCRIPTION
**Pull request by code-elinka**
There are two changes:
* Session Affinity Tracking Policy shouldn't
be updated if the flag for it wasn't set.
* Backend should be updated when
ConnectionTrackingPolicy has changed.

### CODE REVIEW

File: pkg/loadbalancers/l4netlb.go [hunk 1]
```diff
@@ -266,12 +266,13 @@ func (l4netlb *L4NetLB) provideIPv4HealthChecks(nodeNames []string, result *L4Ne
 // connectionTrackingPolicy returns BackendServiceConnectionTrafficPolicy
 // based on StrongSessionAffinity and IdleTimeoutSec
 func (l4netlb *L4NetLB) connectionTrackingPolicy() (*composite.BackendServiceConnectionTrackingPolicy, error) {
-	connectionTrackingPolicy := composite.BackendServiceConnectionTrackingPolicy{}
-	connectionTrackingPolicy.EnableStrongAffinity = annotations.HasStrongSessionAffinityAnnotation(l4netlb.Service)
-	if connectionTrackingPolicy.EnableStrongAffinity {
-		connectionTrackingPolicy.TrackingMode = backends.PerSessionTrackingMode
-		connectionTrackingPolicy.IdleTimeoutSec = int64(*l4netlb.Service.Spec.SessionAffinityConfig.ClientIP.TimeoutSeconds)
+	if !l4netlb.enableStrongSessionAffinity || !annotations.HasStrongSessionAffinityAnnotation(l4netlb.Service) {
```
Comment (aojea) [inline pkg/loadbalancers/l4netlb.go:269]: this may have an unit test covering the regression

Comment (code-elinka) [reply pkg/loadbalancers/l4netlb.go:269]: Sure, thank you.

File: pkg/backends/backends_test.go [hunk 2]
```diff
@@ -108,3 +108,106 @@ func TestEnsureL4BackendService(t *testing.T) {
 		})
 	}
 }
+
+func TestBackendSvcEqual(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/backends/backends_test.go:112]: Sorry, I am usually not a fun of being picky about tests, but as I am substituting Cezary....

I don't think we should have this way of storing old/new in map[string]string

I think there are 2 ways to do it better:

1. Just store oldBackendService newBackendService as structs in testCases
2. You could define those backend services even before testsCases in variables. So it will be:
```

	backendService :=  &composite.BackendService{
		Description:         "description",
		LoadBalancingScheme: "EXTERNAL",
		Protocol:            "TCP",
		SessionAffinity:     "CLIENT_IP",
	}
	backendServiceDesc2 := &composite.BackendService{
		Description:         "different-description",
		LoadBalancingScheme: "EXTERNAL",
		Protocol:            "TCP",
		SessionAffinity:     "CLIENT_IP",
	}
	for _, tc := range []struct {
		desc              string
		oldBackendService *composite.BackendService
		newBackendService *composite.BackendService
                 wantEqual bool
	}{
		{
			desc:              "Test with equal services",
			oldBackendService: backendService,
			newBackendService: backendService,
		},
		{
			desc:              "Test with changed description",
			oldBackendService: backendService,
			newBackendService: backendServiceDesc2,
		},
	},
```

File: pkg/backends/backends_test.go [hunk 3]
```diff
@@ -108,3 +108,106 @@ func TestEnsureL4BackendService(t *testing.T) {
...
+	for _, tc := range []struct {
+		desc                       string
+		backendDesc                map[string]string
+		protocols                  map[string]string
+		affinityTypes              map[string]string
+		schemeTypes                map[string]string
+		strongSessionAffinityFlags map[string]bool
+		idleTimeouts               map[string]int64
+		trackingModes              map[string]string
+		backendsEqual              bool
```
Comment (panslava) [inline pkg/backends/backends_test.go:122]: use `wantEqual`

File: pkg/loadbalancers/l4netlb.go [hunk 4]
```diff
@@ -266,12 +266,13 @@ func (l4netlb *L4NetLB) provideIPv4HealthChecks(nodeNames []string, result *L4Ne
 // connectionTrackingPolicy returns BackendServiceConnectionTrafficPolicy
```
Comment (panslava) [inline pkg/loadbalancers/l4netlb.go:266]: typo: returns BackendServiceConnection*Tracking*Policy

File: pkg/loadbalancers/l4netlb.go [hunk 5]
```diff
@@ -266,12 +266,13 @@ func (l4netlb *L4NetLB) provideIPv4HealthChecks(nodeNames []string, result *L4Ne
...
-	if connectionTrackingPolicy.EnableStrongAffinity {
-		connectionTrackingPolicy.TrackingMode = backends.PerSessionTrackingMode
-		connectionTrackingPolicy.IdleTimeoutSec = int64(*l4netlb.Service.Spec.SessionAffinityConfig.ClientIP.TimeoutSeconds)
+	if !l4netlb.enableStrongSessionAffinity || !annotations.HasStrongSessionAffinityAnnotation(l4netlb.Service) {
+		return nil, nil
 	}
+	connectionTrackingPolicy := composite.BackendServiceConnectionTrackingPolicy{}
+	connectionTrackingPolicy.EnableStrongAffinity = true
+	connectionTrackingPolicy.TrackingMode = backends.PerSessionTrackingMode
+	connectionTrackingPolicy.IdleTimeoutSec = int64(*l4netlb.Service.Spec.SessionAffinityConfig.ClientIP.TimeoutSeconds)
```
Comment (panslava) [inline pkg/loadbalancers/l4netlb.go:275]: Could it happen this will be nil?

File: pkg/utils/utils.go [hunk 6]
```diff
@@ -421,6 +422,20 @@ func EqualResourceIDs(a, b string) bool {
...
+func EqualConnectionTrackingPolicy(a, b *composite.BackendServiceConnectionTrackingPolicy) bool {
+	if a == nil || b == nil {
+		return a == b
+	}
+	return a.TrackingMode == b.TrackingMode &&
+		a.EnableStrongAffinity == b.EnableStrongAffinity &&
+		a.IdleTimeoutSec == b.IdleTimeoutSec &&
+		a.ConnectionPersistenceOnUnhealthyBackends == b.ConnectionPersistenceOnUnhealthyBackends &&
+		EqualStringSets(a.ForceSendFields, b.ForceSendFields) &&
+		EqualStringSets(a.NullFields, b.NullFields)
```
Comment (panslava) [inline pkg/utils/utils.go:436]: Are you sure we need to compare ForceSendFields and NullFields? I personally don't know, and not sure, just never saw this in our code

File: pkg/utils/utils_test.go [hunk 7]
```diff
@@ -1634,3 +1635,113 @@ func TestIsUnsupportedFeatureError(t *testing.T) {
 		})
 	}
 }
+
+func TestEqualConnectionTrackingPolicy(t *testing.T) {
```
Comment (panslava) [inline pkg/utils/utils_test.go:1639]: The same comments as in EqualBackends

**Review by panslava — CHANGES_REQUESTED** [review]

File: pkg/utils/utils.go (hunk 6 above)
Comment (mmamczur) [reply pkg/utils/utils.go:436]: ForceSendFields and NullFields are not really parto of BackendService but rather something that structs generated from protos have. You use them to indicate that you want to send an empty value for a field even if it's not set in a patch request etc.
I think we should not compare them

Comment (code-elinka) [reply pkg/utils/utils.go]: Please check now :)

File: pkg/utils/utils.go [hunk 8]
```diff
@@ -421,6 +422,20 @@ func EqualResourceIDs(a, b string) bool {
 	return aId.Equal(bId)
 }
 
+// EqualConnectionTrackingPolicy returns true if both elements are equal
+// and return false if at least one parameter is different
```
[the remark this answers is not in the document]
Comment (code-elinka) [reply pkg/utils/utils.go]: I moved it but left both tests. Tell me if you insist on removing the test of helper function.

File: pkg/backends/backends.go [hunk 9]
```diff
@@ -33,7 +33,7 @@ import (
 
 const (
 	DefaultConnectionDrainingTimeoutSeconds = 30
```
Comment (cezarygerard) [inline pkg/backends/backends.go:35]: are these used outside of this package?
if not then please rename to lower case

	DefaultConnectionDrainingTimeoutSeconds
	PerSessionTrackingMode

File: pkg/backends/backends_test.go [hunk 10]
```diff
@@ -108,3 +110,296 @@ func TestEnsureL4BackendService(t *testing.T) {
...
+		},
+		{
+			// NullFields change is not supported yet
+			// that's why wantEqual = True. Change this case when the support starts.
+			desc: "NullFields update",
+			oldConnectionTrackingPolicy: &composite.BackendServiceConnectionTrackingPolicy{
+				EnableStrongAffinity: false,
+				IdleTimeoutSec:       defaultIdleTimeout,
+				TrackingMode:         perSessionTrackingMode,
+				NullFields:           []string{"abc"},
```
Comment (cezarygerard) [inline pkg/backends/backends_test.go:384]: please remove this test case

it's confusing ihmo

File: pkg/backends/backends_test.go [hunk 11]
```diff
@@ -108,3 +110,296 @@ func TestEnsureL4BackendService(t *testing.T) {
...
+				ForceSendFields:      []string{"abc"},
+			},
+			newConnectionTrackingPolicy: &composite.BackendServiceConnectionTrackingPolicy{
+				EnableStrongAffinity: false,
+				IdleTimeoutSec:       defaultIdleTimeout,
+				TrackingMode:         perSessionTrackingMode,
+				ForceSendFields:      []string{"xyz"},
+			},
+			wantEqual: true,
+		},
```
Comment (cezarygerard) [inline pkg/backends/backends_test.go:375]: please remove this test case

it's confusing ihmo

File: pkg/backends/backends_test.go [hunk 12]
```diff
@@ -108,3 +110,296 @@ func TestEnsureL4BackendService(t *testing.T) {
...
+				EnableStrongAffinity: true,
+				IdleTimeoutSec:       prolongedIdleTimeout,
+				TrackingMode:         perSessionTrackingMode,
+			},
+			wantEqual: false,
+		},
+		{
+			// ConnectionPersistenceOnUnhealthyBackends change is not supported yet
+			// that's why wantEqual = True. Change this case when the support starts.
+			desc: "Connection Persistence update",
```
Comment (cezarygerard) [inline pkg/backends/backends_test.go:343]: desc: "ConnectionPersistenceOnUnhealthyBackends uodated by customer"

File: pkg/backends/backends_test.go [hunk 13]
```diff
@@ -108,3 +110,296 @@ func TestEnsureL4BackendService(t *testing.T) {
...
+			oldConnectionTrackingPolicy: &composite.BackendServiceConnectionTrackingPolicy{
+				EnableStrongAffinity:                     false,
+				IdleTimeoutSec:                           defaultIdleTimeout,
+				TrackingMode:                             perSessionTrackingMode,
+				ConnectionPersistenceOnUnhealthyBackends: "DEFAULT_FOR_PROTOCOL",
```
Comment (cezarygerard) [inline pkg/backends/backends_test.go:348]: put "non-default" into oldConnectionTrackingPolicy and DEFAULT_FOR_PROTOCOL in the newConnectionTrackingPolicy

File: pkg/backends/backends.go (hunk 9 above)
Comment (code-elinka) [reply pkg/backends/backends.go:35]: I reused them into l4netlb.go + l4lb_test.go. Do you want to have something unique for those places and change these things to private?

File: pkg/backends/backends_test.go (hunk 12 above)
Comment (code-elinka) [reply pkg/backends/backends_test.go]: changed a little. is it better?

File: pkg/backends/backends_test.go [hunk 14]
```diff
@@ -108,3 +110,260 @@ func TestEnsureL4BackendService(t *testing.T) {
...
+			t.Parallel()
+			if res := backendSvcEqual(tc.oldBackendService, tc.newBackendService) == tc.wantEqual; !res {
+				t.Errorf("backendSvcEqual() returned %v, expected %v. Diff(oldScv, newSvc): %s",
+					res, tc.wantEqual, cmp.Diff(tc.oldBackendService, tc.newBackendService))
+			}
+		})
+	}
+}
+
+func TestConnectionTrackingPolicyEqual(t *testing.T) {
```
Comment (cezarygerard) [inline pkg/backends/backends_test.go:263]: please drop this test and move test cases to TestBackendSvcEqual if they are not present yet

File: pkg/backends/backends_test.go [hunk 15]
```diff
@@ -108,3 +110,260 @@ func TestEnsureL4BackendService(t *testing.T) {
...
+				Description:         "same_description",
+				Protocol:            "TCP",
+				SessionAffinity:     string(v1.ServiceAffinityClientIP),
+				LoadBalancingScheme: string(cloud.SchemeExternal),
+				ConnectionTrackingPolicy: &composite.BackendServiceConnectionTrackingPolicy{
+					EnableStrongAffinity: false,
+					IdleTimeoutSec:       prolongedIdleTimeout,
+					TrackingMode:         defaultTrackingMode,
+				},
+			},
```
Comment (cezarygerard) [inline pkg/backends/backends_test.go:159]: IMHO it is better not to repeat the same fields between oldBackendService and newBackendService

only initialize fields that differ for readability sake
e.g.

```
		{
			desc: "Test with changed idle timeout",
			oldBackendService: &composite.BackendService{
				ConnectionTrackingPolicy: &composite.BackendServiceConnectionTrackingPolicy{
					IdleTimeoutSec: prolongedIdleTimeout,
				},
			},
			newBackendService: &composite.BackendService{
				ConnectionTrackingPolicy: &composite.BackendServiceConnectionTrackingPolicy{
					IdleTimeoutSec: defaultIdleTimeout,
				},
			},
			wantEqual: false,
		},
		
```

you can also add one more test case 
```
		{
			desc:              "Test empty backedns are equal",
			oldBackendService: &composite.BackendService{},
			newBackendService: &composite.BackendService{},
			wantEqual:         true,
		},
```

to also test the null safety

File: pkg/backends/backends_test.go [hunk 16]
```diff
@@ -108,3 +110,133 @@ func TestEnsureL4BackendService(t *testing.T) {
 		})
 	}
 }
+
+// TestBackendSvcEqual checks that backendSvcEqual() and
+// connectionTrackingPolicyEqual() (as a part ofit  backendSvcEqual)
+// return expected results for two resources compared.
+func TestBackendSvcEqual(t *testing.T) {
```
Comment (cezarygerard) [inline pkg/backends/backends_test.go:117]: this test does not assert on
- TrackingMode
- Network
- HealthChecks

Comment (code-elinka) [reply pkg/backends/backends_test.go:117]: done
